src/value.h

```diff
@@ -230,7 +247,7 @@ class Value
     return stream;
   }
 
-  typedef boost::variant< boost::blank, bool, double, str_utf8_wrapper, VectorType, RangeType > Variant;
+  typedef boost::variant< boost::blank, bool, double, str_utf8_wrapper, VectorType, RangeType, std::shared_ptr<Expression>> Variant;
```

Comment:
I have to think about this, but at first read, it feels odd: We're storing an Expression, which is an AST construct, in a value, which is a run-time construct. For all other types, it's the other way around: The expression embeds a literal value.

My immediate hunch would be to define a `FunctionLiteral` stored in a value and aggregate that into the `FunctionDefinition` expression.

We don't currently need a good separation between the AST and the run-time, but this makes the coupling a bit tight.

-> could be refactored later though

Reply:
Oh, and as a side-question: Why store an expression and not the `FunctionDefinition` directly?

Reply:
Agreed, the evaluate to a copy of itself seems a bit odd too.

Reply:
Did you have a chance to think about this? I think this is the last remaining concern I have about this PR..

Reply:
I've now added a `FunctionType` for use as value. It still uses `Expression` and `AssignmentList` though.